### PR TITLE
internals: minor clean-up of `types` ordinal procs

### PR DESCRIPTION
**Pull request by saem**
## Summary

Minor changes to ordinal related routines in the `types` module. These
are in preparation for a larget `semCase` refactor to use `nkError`.

## Details

Minor changes `firstOrd` and `lastOrd`, largely removing vestigial code
for when enum inheritance was a feature. Replaced bare `localReports`
usage with internal asserts/errors.

Unfortunately, the `ConfigRef` requirement for `firstOrd`, `lastOrd`,
and `lengthOrd` couldn't be dropped as it needs to look at
`ConfigRef.intSize` for `tyInt`. Ordinal querying happens all over the
compiler, with a number of subtle assumptions. Reworking these to use
the extra range of the `Int128` return type wasn't possible, yet.

I also discovered that enums could inherit from other enums, there is
still some vestigial code for this; will address in a follow-up PR.

---

## Notes for Reviewers
I pushed the two float related procs out of the way, it was too annoying,
unfrotunately git diff is having a really hard time with it. No actual
changes to `lengthOrd` or the float routines.

### CODE REVIEW

File: compiler/ast/types.nim [hunk 1]
```diff
@@ -443,49 +449,29 @@ proc firstOrd*(conf: ConfigRef; t: PType): Int128 =
...
      tyStatic, tyInferred, tyUserTypeClasses, tyLent:
     result = firstOrd(conf, lastSon(t))
   of tyOrdinal:
-    if t.len > 0: result = firstOrd(conf, lastSon(t))
-    else:
-      conf.localReport InternalReport(
-        kind: rintUnreachable,
-        msg: "invalid kind for firstOrd(" & $t.kind & ')')
-
+    # TODO: convert to an invalid return and have callee handle it
```
Comment (zerbina) [inline compiler/ast/types.nim]: Did you mean `caller` instead of `callee` here?

File: compiler/ast/types.nim [hunk 2]
```diff
@@ -443,49 +449,29 @@ proc firstOrd*(conf: ConfigRef; t: PType): Int128 =
...
+    conf.internalAssert(t.len > 0, "firstOrd - abstract ordinal type given")
+    result = firstOrd(conf, lastSon(t))
   of tyUncheckedArray, tyCstring:
     result = Zero
   else:
-    conf.localReport InternalReport(
-      kind: rintUnreachable,
-      msg: "invalid kind for firstOrd(" & $t.kind & ')')
+    conf.internalError("firstOrd - non-ordinal type given: " & $t.kind)
```
Comment (zerbina) [inline compiler/ast/types.nim]: I think an `unreachable` would fit better here, but I'm also okay with leaving the `internalError`.

File: compiler/ast/types.nim [hunk 3]
```diff
@@ -513,27 +499,52 @@ proc lastOrd*(conf: ConfigRef; t: PType): Int128 =
...
+        assert(t.n[^1].kind == nkSym)
+        toInt128(t.n[^1].sym.position)
+      else:
+        # currently empty enums aren't support as a type, but they should be
+        Zero
   of tyGenericInst, tyDistinct, tyTypeDesc, tyAlias, tySink,
      tyStatic, tyInferred, tyUserTypeClasses, tyLent:
     result = lastOrd(conf, lastSon(t))
-  of tyProxy: result = Zero
+  of tyProxy: result = Zero # xxx: this seems off; also not in `firstOrd`
```
Comment (zerbina) [inline compiler/ast/types.nim]: Given that `tyProxy` is only used for errors now, I think this should be handled as part of the `else` branch (i.e. treated as unreachable).

Adjusting it doesn't have to happen as part of this PR, though.

File: compiler/ast/types.nim [hunk 4]
```diff
@@ -513,27 +499,52 @@ proc lastOrd*(conf: ConfigRef; t: PType): Int128 =
...
   of tyOrdinal:
-    if t.len > 0: result = lastOrd(conf, lastSon(t))
-    else:
-      conf.localReport InternalReport(
-        kind: rintUnreachable,
-        msg: "invalid kind for firstOrd(" & $t.kind & ')')
-  of tyUncheckedArray:
+    conf.internalAssert(t.len > 0, "lastOrd - abstract ordinal type given")
+    result = lastOrd(conf, lastSon(t))
+  of tyUncheckedArray: # xxx: what about `tyCstring`; see `firstOrd`
```
Comment (zerbina) [inline compiler/ast/types.nim]: Neither `cstring` nor `UncheckedArray` have a compile-time-known lower or upper bound, so I think both shouldn't be handled by `firstOrd`/`lastOrd`.

This is also not something I expect to happen now.

File: compiler/ast/types.nim [hunk 5]
```diff
@@ -513,27 +499,52 @@ proc lastOrd*(conf: ConfigRef; t: PType): Int128 =
...
   of tyEnum:
-    if t.n.len > 0:
-      assert(t.n[^1].kind == nkSym)
-      result = toInt128(t.n[^1].sym.position)
+    result =
+      if t.n.len > 0:
+        assert(t.n[^1].kind == nkSym)
+        toInt128(t.n[^1].sym.position)
+      else:
+        # currently empty enums aren't support as a type, but they should be
```
Comment (zerbina) [inline compiler/ast/types.nim]:
```suggestion
        # currently empty enums aren't supported as a type, but they should be
```

**Review by zerbina — COMMENTED** [review]
I left a few small suggestions and comments, but nothing that's a hard blocker.

Regarding using the return value to indicate errors, I don't think doing so is necessary. Calling `firstOrd` or `lastOrd` with a type that doesn't have compile-time-known limits hints at a logic error somewhere and, in my opinion, should abort compilation.

File: compiler/ast/types.nim (hunk 1 above)
Comment (saem) [reply compiler/ast/types.nim]: I absolutely do, thanks

File: compiler/ast/types.nim (hunk 2 above)
Comment (saem) [reply compiler/ast/types.nim]: I want to crash the compiler and force bug fixes + remove open reports code, I could use the idiom I suppose.

File: compiler/ast/types.nim (hunk 3 above)
Comment (saem) [reply compiler/ast/types.nim]: I was thinking `tyProxy` branches are where we might add in forgiveness for suggest situations, while everything else is a programming error.

File: compiler/ast/types.nim (hunk 4 above)
Comment (saem) [reply compiler/ast/types.nim]: Yeah they're pessimisticly `-1`, as in empty; unless we have some hints like it's not `nil`.

File: compiler/ast/types.nim (hunk 5 above)
Comment (saem) [reply compiler/ast/types.nim]: Thank you, thank you.

File: compiler/ast/types.nim (hunk 1 above)
Comment (saem) [reply compiler/ast/types.nim]: ultimately removed the comment as it should simply be a failure.

File: compiler/ast/types.nim (hunk 3 above)
Comment (saem) [reply compiler/ast/types.nim]: Captured the above via comment.

File: compiler/ast/types.nim (hunk 4 above)
Comment (saem) [reply compiler/ast/types.nim]: captured the above as a comment; even if somewhat oblique.

**Review by zerbina — APPROVED** [review]

**Pull request merged**